src/coprocessor_v2/plugin_registry.rs

```diff
@@ -98,28 +313,102 @@ impl CoprocessorPlugin for LoadedPlugin {
+//        let library_path = pkgname_to_libname("example-plugin");
+//        let loaded_plugin = unsafe { LoadedPlugin::new(&library_path).unwrap() };
 //
-//        assert_eq!(plugin_name, "example-plugin");
+//        assert_eq!(loaded_plugin.name(), "example-plugin");
+//        assert_eq!(loaded_plugin.file_name().to_str().unwrap(), library_path);
 //    }
 //
 //    #[test]
-//    fn plugin_registry_load_and_get_plugin() {
-//        let mut plugin_registry = PluginRegistry::default();
-//        let plugin_name = plugin_registry
-//            .load_plugin(pkgname_to_libname("example-plugin"))
-//            .unwrap();
-//        let plugin = plugin_registry.get_plugin(plugin_name).unwrap();
+//    fn registry_load_and_get_plugin() {
+//        let registry = PluginRegistry::new();
+//        let library_path = pkgname_to_libname("example-plugin");
+//        let plugin_name = registry.load_plugin(&library_path).unwrap();
+//
+//        let plugin = registry.get_plugin(plugin_name).unwrap();
 //
 //        assert_eq!(plugin.name(), "example-plugin");
+//        assert_eq!(registry.loaded_plugin_names(), vec!["example-plugin"]);
+//    }
+//
+//    #[test]
+//    fn registry_unload_plugin_by_name() {
+//        let registry = PluginRegistry::new();
+//        let library_path = pkgname_to_libname("example-plugin");
+//        let plugin_name = registry.load_plugin(&library_path).unwrap();
+//
+//        assert!(registry.get_plugin(plugin_name).is_some());
+//
+//        registry.unload_plugin_by_name(plugin_name);
+//
+//        assert!(registry.get_plugin(plugin_name).is_none());
+//        assert_eq!(registry.loaded_plugin_names().len(), 0);
+//    }
+//
+//    #[test]
+//    fn plugin_registry_unload_plugin_by_path() {
+//        let registry = PluginRegistry::new();
+//        let library_path = pkgname_to_libname("example-plugin");
+//        let plugin_name = registry.load_plugin(&library_path).unwrap();
+//
+//        assert!(registry.get_plugin(plugin_name).is_some());
+//
+//        registry.unload_plugin_by_path(library_path);
+//
+//        assert!(registry.get_plugin(plugin_name).is_none());
+//        assert_eq!(registry.loaded_plugin_names().len(), 0);
+//    }
+//
+//    #[test]
+//    fn plugin_registry_hot_reloading() {
+//        let build_dir = std::env::current_exe()
+//            .map(|p| p.as_path().parent().unwrap().to_owned())
+//            .unwrap();
+//        let coprocessor_dir = build_dir.join("coprocessors");
+//
+//        std::fs::create_dir_all(&coprocessor_dir).unwrap();
+//        let library_path = pkgname_to_libname("example-plugin");
+//
+//        let mut registry = PluginRegistry::new();
+//        registry.start_hot_reloading(&coprocessor_dir).unwrap();
+//
+//        // trigger loading
+//        std::fs::copy(
+//            build_dir.join(&library_path),
+//            coprocessor_dir.join(&library_path),
+//        )
+//        .unwrap();
+//        std::thread::sleep(Duration::from_secs(4));
```

Comment:
From my experience, tests involving non-deterministic behaviors (like waiting for x seconds) have a high possibility of random failing on CI. We could see if this test works later when we run this on CI.

Reply:
Yeah, I thought the same thing :see_no_evil: 
But on the other hand I think we can't really get around it? Because the file system watcher debounces events with a delay of 3s...

Reply:
Maybe we _could_ implement the [`Watcher`](https://docs.rs/notify/4.0.15/notify/trait.Watcher.html) with our own watcher where we can manually trigger events? Maybe that's a good way to go?

Reply:
We could later decide on how to handle this. If this case randomly fails, the bot would automatically open up an issue.